### PR TITLE
Review SCSCF UT changes from SDM refactor

### PR DESCRIPTION
**Pull request by AlexHockey**
@KristaHyer I'm going to review your changes in this PR as the other one is laggy AF because it's too large.

### CODE REVIEW

File: src/ut/scscf_test.cpp [hunk 1]
```diff
@@ -407,30 +467,25 @@ class SCSCFTest : public SCSCFTestBase
...
+
+  // Expect calls to look up the iFCs and bindings of the callee.
+  HSSConnection::irs_info irs_info_2;
+  Bindings bindings;
+  setup_callee_info(irs_info_2, bindings);
+  std::vector<std::string> ifc_list_2;
+  add_ifc_info(ifc_list_2, 1, {"<Method>INVITE</Method>", "<SessionCase>1</SessionCase><!-- terminating-registered -->"}, "sip:5.2.3.4:56787;transport=UDP");
+  add_ifc_info(ifc_list_2, 2, {"<Method>QWERTY_UIOP</Method>"}, "sip:sholes.example.com");
+  add_ifc_info(ifc_list_2, 3, {"<Method>INVITE</Method>"}, "sip:6.2.3.4:56786;transport=UDP");
+  char* ifc_str_2 = add_ifcs(irs_info_2, ifc_list_2, "sip:6505551234@homedomain");
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: @sathiyan-sivathas you had a better way of doing this. When you have merged your testing common code can you prod @KristaHyer. @KristaHyer when this is done you can remove the return value from `add_ifcs` and the world will be very very slightly better.

File: src/ut/scscf_test.cpp [hunk 2]
```diff
@@ -1105,10 +1162,230 @@ void SCSCFTestBase::doSlowFailureFlow(SCSCFMessage& msg,
...
+  AssociatedURIs associated_uris = {};
+  associated_uris.add_uri(uri, barred);
+  irs_info._associated_uris = associated_uris;
+
+  irs_info._regstate = RegDataXMLUtils::STATE_REGISTERED;
+  irs_info._prev_regstate = "";
+
+  std::map<std::string, Ifcs> service_profiles;
+  Ifcs ifcs;
+  service_profiles.insert(std::make_pair("first_key" , ifcs));
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: What does "first_key" mean?

File: src/ut/scscf_test.cpp [hunk 3]
```diff
@@ -1105,10 +1162,230 @@ void SCSCFTestBase::doSlowFailureFlow(SCSCFMessage& msg,
...
+                                  std::string contact,
+                                  bool emergency)
+{
+  Binding* binding = AoRTestUtils::build_binding(uri,
+                                                 time(NULL),
+                                                 contact);
+  if (emergency)
+  {
+    binding->_emergency_registration = true;
+  }
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: Why not just `binding->_emergency_registration = emergency;`? ;-)

File: src/ut/scscf_test.cpp (hunk 2 above)
Comment (AlexHockey) [reply src/ut/scscf_test.cpp]: Ahh I see. Can we call this "placeholder" instead, make it into a constant, and add a comment explaining what is going on?

File: src/ut/scscf_test.cpp [hunk 4]
```diff
@@ -1105,10 +1162,230 @@ void SCSCFTestBase::doSlowFailureFlow(SCSCFMessage& msg,
...
+  Ifcs ifcs = Ifcs(root, root->first_node("ServiceProfile"), NULL, 0);
+  irs_info._service_profiles.insert(std::make_pair(uri, ifcs));
+
+  // If blank iFC (used as placeholder for tests where iFCs are not required) is
+  // present, remove it.
+  std::map<std::string, Ifcs>::iterator searching = irs_info._service_profiles.find("first_key");
+  if (searching != irs_info._service_profiles.end())
+  {
+    irs_info._service_profiles.erase("first_key");
+  }
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: I think you can just do `irs_info._service_profiles.erase("first_key");` without checking first.

File: src/ut/scscf_test.cpp [hunk 5]
```diff
@@ -1105,10 +1162,230 @@ void SCSCFTestBase::doSlowFailureFlow(SCSCFMessage& msg,
...
+}
+
+// Set up EXPECT_CALL(s) to get the subscriber state for a subscriber.
+// There should always be 1 call to get the subscriber state (to find the iFCs),
+// and then an additional call for each AS that returns the INVITE to the S-CSCF
+// (as a check will take place to see if a retarget has happened.
+void SCSCFTestBase::expect_get_subscriber_state(HSSConnection::irs_info& irs_info,
+                                                std::string uri,
+                                                bool check_retarget,
+                                                int num_matching_ifcs)
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: Why not just have one parameter (`times`, defaults to 1) which is the number of times to expect the call?

File: src/ut/scscf_test.cpp [hunk 6]
```diff
@@ -1105,10 +1162,230 @@ void SCSCFTestBase::doSlowFailureFlow(SCSCFMessage& msg,
...
+// Set up the EXPECT_CALLs relating to the standard callee.
+// There should be one call to get the iFCs (in the subscriber state), and one
+// to get the bindings. Also, for every matched on AS, there should be an
+// additional call to get the subscriber state, to check that the returned
+// INVITE has not been retargeted.
+void SCSCFTestBase::expect_get_callee_info(HSSConnection::irs_info& irs_info,
+                                           Bindings& bindings,
+                                           std::string uri,
+                                           bool check_retarget,
+                                           int num_matching_ifcs)
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: Ditto.

File: src/ut/scscf_test.cpp [hunk 7]
```diff
@@ -1118,54 +1395,85 @@ TEST_F(SCSCFTest, TestSimpleMainline)
...
+  Bindings bindings;
+  setup_callee_info(irs_info, bindings);
+  EXPECT_CALL(*_sm, get_subscriber_state(IrsQueryWithPublicId("sip:6505551234@homedomain"), _, _))
+    .Times(2)
+    .WillRepeatedly(DoAll(SetArgReferee<1>(irs_info),
+                          Return(HTTP_OK)));
+  EXPECT_CALL(*_sm, get_bindings("sip:6505551234@homedomain", _, _))
+    .WillOnce(DoAll(SetArgReferee<1>(bindings),
+                    Return(HTTP_OK)))
+    .RetiresOnSaturation();
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: Couldn't you just have one `EXPECT_CALL` and two `WillOnce` calls?

File: src/ut/scscf_test.cpp [hunk 8]
```diff
@@ -1885,25 +1886,57 @@ TEST_F(SCSCFTest, TestNoEnumWhenGRUU)
...
+  // Set up callee info.
+  HSSConnection::irs_info irs_info_2;
+  setup_irs_info(irs_info_2, "+15108580271", "homedomain");
+  Bindings bindings;
+  Binding* binding = AoRTestUtils::build_binding("sip:+15108580271@homedomain",
+                                                 time(NULL),
+                                                 "sip:wuntootreefower@10.114.61.213:5061;transport=tcp;ob",
+                                                 30);
+  binding->_params["+sip.instance"] = "abcde";
+  bindings.insert(std::make_pair("sip:wuntootreefower@10.114.61.213:5061;transport=tcp;ob", binding));
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: Would it be easier here (and in other places) to call setup_bindings and then just do `bindings.begin()->_params = ["+sip.instance"] = "abcde";`. That guarantees we're setting the binding up the same way as normal except for the new param.

File: src/ut/scscf_test.cpp [hunk 9]
```diff
@@ -6102,12 +6019,18 @@ void SCSCFTestBase::doAsOriginated(const std::string& msg, bool expect_orig)
...
 
   // Inject succesful responses to finish up the flow
   inject_msg(respond_to_current_txdata(200));
   inject_msg(respond_to_current_txdata(200));
   inject_msg(respond_to_current_txdata(200));
+
+  if (ifc_str_1 != NULL)
+  {
+    free(ifc_str_1);
+  }
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: FYI you never need to do this. Calling free or delete on a NULL pointer is safe.

File: src/ut/scscf_test.cpp [hunk 10]
```diff
@@ -9426,25 +9485,41 @@ TEST_F(SCSCFTest, HSSTimeoutOnPutRegData)
...
+  // First call succeeds.
+  EXPECT_CALL(*_sm, get_subscriber_state(IrsQueryWithPublicId("sip:6505551234@homedomain"), _, _))
+    .WillOnce(DoAll(SetArgReferee<1>(irs_info),
+                    Return(HTTP_OK)))
+    .RetiresOnSaturation();
+  // Second and third calls fail.
+  EXPECT_CALL(*_sm, get_subscriber_state(IrsQueryWithPublicId("sip:6505551234@homedomain"), _, _))
+    .Times(2)
+    .WillRepeatedly(Return(HTTP_SERVER_UNAVAILABLE))
+    .RetiresOnSaturation();
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: Again, I would have thought you could just have one EXPECT_CALL and multiple WillOnce.

File: src/ut/scscf_test.cpp [hunk 11]
```diff
@@ -9509,30 +9578,53 @@ TEST_F(SCSCFTest, HSSTimeoutOnCdiv)
...
+  // First call succeeds.
+  EXPECT_CALL(*_sm, get_subscriber_state(IrsQueryWithPublicId("sip:6505551234@homedomain"), _, _))
+    .WillOnce(DoAll(SetArgReferee<1>(irs_info),
+                    Return(HTTP_OK)))
+    .RetiresOnSaturation();
+  // Second and third calls fail.
+  EXPECT_CALL(*_sm, get_subscriber_state(IrsQueryWithPublicId("sip:6505551234@homedomain"), _, _))
+    .Times(2)
+    .WillRepeatedly(Return(HTTP_NOT_FOUND))
+    .RetiresOnSaturation();
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: Ditto

File: src/ut/scscf_test.cpp [hunk 12]
```diff
@@ -9697,89 +9724,113 @@ TEST_F(SCSCFTest, TestCallerNotBarred)
...
+  // registrations after it is realised the served user is barred, and one to
+  // find the target to route to.
+  EXPECT_CALL(*_sm, get_bindings("sip:6505551234@homedomain", _, _))
+    .WillOnce(DoAll(SetArgReferee<1>(bindings),
+                    Return(HTTP_OK)))
+    .RetiresOnSaturation();
+  EXPECT_CALL(*_sm, get_bindings("sip:6505551234@homedomain", _, _))
+    .WillOnce(DoAll(SetArgReferee<1>(bindings_2),
+                    Return(HTTP_OK)))
+    .RetiresOnSaturation();
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: Again. I'm going to stop commenting on these, but ideally you would remove all `RetiresOnSaturation` calls.

File: src/ut/scscf_test.cpp [hunk 13]
```diff
@@ -10201,10 +10259,24 @@ TEST_F(SCSCFTest, MixedRealAndDummyApplicationServer)
...
 
-// Test that a MESSAGE containing "urn:service:service" in the Request URI is
+
+// Test that a MESSAGE containing "urn:service:sos" in the Request URI is
 // handled.
+//
+// The flow here is non-spec.
+// Example use case - a barred user makes an emergency call, and the call goes
+// through the E-CSCF, not the I-CSCF and S-CSCF. The user then accidentally
+// hangs up, but the call can't be returned since they are barred.
```
Comment (AlexHockey) [inline src/ut/scscf_test.cpp]: I get that you mean AS barring here, but I immediately thought of S-CSCF barring. Can you be more explicit?

**Comment by KristaHyer** [comment]
Closing since this PR was just created to allow easy review

**Pull request closed without merging**